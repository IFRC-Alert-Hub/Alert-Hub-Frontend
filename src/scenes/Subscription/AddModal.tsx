import { Box, Modal, Typography } from "@mui/material";

const style = {
  position: "absolute" as "absolute",
  top: "30%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  width: "40%",
  minWidth: "350px",
  bgcolor: "background.paper",
  boxShadow: 24,
  p: 4,
  borderRadius: "5px",
};

type PropsType = {
  open: boolean;
  handleClose: React.Dispatch<React.SetStateAction<boolean>>;
};

const AddModal = ({ open, handleClose }: PropsType) => {
  return (
    <Modal open={open} onClose={handleClose} aria-labelledby="modal-title">
      <Box sx={style}>
        <Typography id="modal-title" variant="h6" fontWeight={"bold"}>
          Add New Subscription
        </Typography>
        <Box>form content</Box>
      </Box>
    </Modal>
  );
};

export default AddModal;
