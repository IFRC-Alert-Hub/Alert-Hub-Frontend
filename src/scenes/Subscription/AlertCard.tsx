import { Box, Button, Card, CardContent, Typography } from "@mui/material";

type PropsType = {
  title: string;
};

const AlertCard = ({ title }: PropsType) => {
  return (
    <Card sx={{ flex: "0 0 300px", mr: 2, mb: 2 }}>
      <CardContent sx={{ display: "flex", justifyContent: "space-between" }}>
        <Box>
          <Typography
            gutterBottom
            variant="h5"
            component="div"
            sx={{ fontWeight: 600 }}
          >
            {title}
          </Typography>
          <Typography variant="body2">Current Alerts: 10</Typography>
          <Typography variant="body2" color="text.secondary">
            Last updated: July 8, 2023
          </Typography>
        </Box>
        <Box>
          <Button
            variant="text"
            disableRipple
            sx={{
              color: "red",
              minWidth: 0,
              textTransform: "capitalize",
              p: "0 5px 0 0",
              fontSize: "0.875rem",
              "&:hover": {
                opacity: "0.7",
              },
            }}
          >
            View
          </Button>
        </Box>
      </CardContent>
    </Card>
  );
};

export default AlertCard;
