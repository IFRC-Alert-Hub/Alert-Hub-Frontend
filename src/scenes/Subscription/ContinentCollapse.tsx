import { Box, Collapse, IconButton, Typography } from "@mui/material";
import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";
import KeyboardArrowUpIcon from "@mui/icons-material/KeyboardArrowUp";
import { ChangeEvent, useState } from "react";
import {
  CountryType,
  SubscriptionForm,
} from "../../API/queries/getSubscriptions";

interface PropsType {
  continent: string;
  countries: CountryType[];
  selectedRow: SubscriptionForm;
  handleChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const ContinentCollapse = ({
  continent,
  countries,
  selectedRow,
  handleChange,
}: PropsType) => {
  const [open, setOpen] = useState(false);

  return (
    <Box>
      <div className="subs-form-continent">
        <IconButton
          aria-label="expand row"
          size="small"
          onClick={() => setOpen(!open)}
        >
          {open ? <KeyboardArrowUpIcon /> : <KeyboardArrowDownIcon />}
        </IconButton>
        <Typography variant="h6">{continent}</Typography>
        {/* <Box ml={2} sx={{ color: "gray", fontSize: "0.5em" }}>
          {selectedRow.countries.length}/{countries.length} selected
        </Box> */}
      </div>
      <Collapse in={open} sx={{ ml: 4, mr: 4 }}>
        <Box
          display={"flex"}
          flexDirection={"row"}
          flexWrap={"wrap"}
          width={"100%"}
          sx={{
            maxHeight: "200px",
            overflow: "auto",
            border: "1px solid rgb(231, 235, 240)",
            borderRadius: "5px",
          }}
        >
          {countries?.map((country) => (
            <div key={country.id}>
              <input
                className="country-input"
                type="checkbox"
                name="countries"
                id={country.name}
                value={country.id}
                checked={selectedRow["countries"].includes(`${country.id}`)}
                onChange={handleChange}
              />
              <label htmlFor={country.name} className="country-checkbox">
                {country.name}
              </label>
            </div>
          ))}
        </Box>
      </Collapse>
    </Box>
  );
};

export default ContinentCollapse;
