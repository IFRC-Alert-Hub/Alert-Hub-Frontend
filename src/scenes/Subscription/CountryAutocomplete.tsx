import {
  Autocomplete,
  Box,
  FormControl,
  FormHelperText,
  FormLabel,
  TextField,
} from "@mui/material";
import {
  CountryType,
  SubscriptionForm,
} from "../../API/queries/getSubscriptions";

type optionsType = {
  id: string;
  name: string;
  firstLetter: string;
}[];

interface FormErrors {
  [key: string]: boolean;
}

type PropsType = {
  verifyForm: boolean;
  formErrors: FormErrors;
  countryList: CountryType[];
  selectedRow: SubscriptionForm;
  setSelectedRow: React.Dispatch<React.SetStateAction<SubscriptionForm>>;
};

const CountryAutocomplete = ({
  verifyForm,
  formErrors,
  countryList,
  selectedRow,
  setSelectedRow,
}: PropsType) => {
  const options = countryList.map((option) => {
    const firstLetter = option.name[0].toUpperCase();
    return {
      firstLetter: /[0-9]/.test(firstLetter) ? "0-9" : firstLetter,
      ...option,
    };
  });

  const displayValues = options.filter((item) =>
    selectedRow.countries.includes(item.id)
  );

  const handleChange = (event: any, newValue: optionsType) => {
    const countryIds = newValue?.map((item) => item.id);
    setSelectedRow((prevState) => ({
      ...prevState,
      countries: countryIds,
    }));
  };

  return (
    <FormControl
      required
      error={verifyForm && formErrors["countries"]}
      component="fieldset"
      sx={{ width: "100%" }}
    >
      <Box display="flex" sx={{ alignItems: "center" }}>
        <FormLabel className="subs-form-legend">Countries</FormLabel>
        <Box ml={2} sx={{ color: "gray", fontSize: "0.5em" }}>
          {selectedRow.countries.length}/{countryList.length} selected
        </Box>
      </Box>
      <Autocomplete
        multiple
        id="tags-outlined"
        size="small"
        options={options.sort(
          (a, b) => -b.firstLetter.localeCompare(a.firstLetter)
        )}
        groupBy={(option) => option.firstLetter}
        getOptionLabel={(option) => option.name}
        value={displayValues}
        filterSelectedOptions
        onChange={handleChange}
        renderInput={(params) => (
          <TextField {...params} placeholder="Countries" />
        )}
        sx={{ mt: 1, mb: 1 }}
      />
      <FormHelperText>
        {verifyForm &&
          formErrors["countries"] &&
          "You need to select at least one"}
      </FormHelperText>
    </FormControl>
  );
};

export default CountryAutocomplete;
