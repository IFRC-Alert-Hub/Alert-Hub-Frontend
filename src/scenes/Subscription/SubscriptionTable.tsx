import { useEffect, useMemo, useState } from "react";
import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Paper,
  TablePagination,
  Button,
  Tooltip,
  Box,
  Typography,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogContentText,
  DialogActions,
} from "@mui/material";
import {
  ContinentType,
  CountryType,
  GET_SUBSCRIPTIONS,
  SubscriptionForm,
  SubscriptionItem,
} from "../../API/queries/getSubscriptions";
import { useMutation } from "@apollo/client";
import { DELETE_SUBSCRIPTION } from "../../API/mutations/subscriptionMutation";
import { subscription_module } from "../../API/API_Links";
import ModalForm from "./ModalForm";

type PropsType = {
  selectedRow: SubscriptionForm;
  setSelectedRow: React.Dispatch<React.SetStateAction<SubscriptionForm>>;
  formType: string;
  countryList: CountryType[];
  regionList: ContinentType[];
  tableData: SubscriptionItem[];
  setTableData: React.Dispatch<React.SetStateAction<SubscriptionItem[]>>;
  open: boolean;
  handleOpen: (type: string, id?: string) => void;
  handleClose: () => void;
};

interface UpdatedRow extends SubscriptionItem {
  filteredCountryNames: (string | undefined)[];
}

const SubscriptionTable = ({
  selectedRow,
  setSelectedRow,
  formType,
  countryList,
  regionList,
  tableData,
  setTableData,
  open,
  handleOpen,
  handleClose,
}: PropsType) => {
  const [deleteSubscription] = useMutation(DELETE_SUBSCRIPTION, {
    refetchQueries: [GET_SUBSCRIPTIONS, "FetchSubscriptions"],
    client: subscription_module,
  });
  const [page, setPage] = useState(0);
  const [rowsPerPage, setRowsPerPage] = useState(5);
  const [updatedRows, setUpdatedRows] = useState<UpdatedRow[]>();
  const [confirmDelete, setConfirmDelete] = useState(false);

  // show the first three countries
  useEffect(() => {
    const updatedItems = tableData.map((row: SubscriptionItem) => {
      if (row.countryIds?.length > 3) {
        const filteredCountryIds = row.countryIds.slice(0, 3);
        const filteredCountryNames = filteredCountryIds.map((countryId) => {
          const foundCountry = countryList.find(
            (country) => country.id === countryId.toString()
          );
          return foundCountry?.name;
        });
        filteredCountryNames.push("...");
        return { ...row, filteredCountryNames };
      } else {
        const filteredCountryIds = row.countryIds;
        const filteredCountryNames = filteredCountryIds.map((countryId) => {
          const foundCountry = countryList?.find(
            (country) => country.id === countryId.toString()
          );
          return foundCountry?.name;
        });
        return { ...row, filteredCountryNames };
      }
    });
    setUpdatedRows(updatedItems);
  }, [tableData, countryList]);

  // Avoid a layout jump when reaching the last page with empty rows.
  const emptyRows = useMemo(
    () =>
      page > 0 ? Math.max(0, (1 + page) * rowsPerPage - tableData?.length) : 0,
    [page, rowsPerPage, tableData]
  );
  const visibleRows = useMemo(
    () =>
      updatedRows?.slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage),
    [page, rowsPerPage, updatedRows]
  );

  const handleChangePage = (
    event: React.MouseEvent<HTMLButtonElement> | null,
    newPage: number
  ) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setRowsPerPage(parseInt(event.target.value, 10));
    setPage(0);
  };

  const handleDelete = (id: string) => {
    deleteSubscription({ variables: { subscriptionId: parseInt(id) } });
    setTableData(tableData.filter((item) => item.id !== id));
  };

  return (
    <TableContainer component={Paper}>
      <Table sx={{ minWidth: 650 }} aria-label="subscription table">
        <TableHead>
          <TableRow>
            <TableCell
              align="center"
              sx={{
                fontSize: "0.875rem",
                fontWeight: "600",
                minWidth: "150px",
              }}
            >
              Title
            </TableCell>
            <TableCell
              align="center"
              sx={{ fontSize: "0.875rem", fontWeight: "600" }}
            >
              Subscribed Countries
            </TableCell>
            <TableCell
              align="center"
              sx={{ fontSize: "0.875rem", fontWeight: "600" }}
            >
              Urgency
            </TableCell>
            <TableCell
              align="center"
              sx={{ fontSize: "0.875rem", fontWeight: "600" }}
            >
              Severity
            </TableCell>
            <TableCell
              align="center"
              sx={{ fontSize: "0.875rem", fontWeight: "600" }}
            >
              Certainty
            </TableCell>
            <TableCell
              align="center"
              sx={{ fontSize: "0.875rem", fontWeight: "600" }}
            >
              Notification Methods
            </TableCell>
            <TableCell
              align="center"
              sx={{ fontSize: "0.875rem", fontWeight: "600" }}
            >
              Operations
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {visibleRows?.map((row: UpdatedRow) => (
            <TableRow key={row.id}>
              <TableCell
                align="center"
                sx={{ fontSize: "0.875rem" }}
                component="th"
                scope="row"
              >
                <Button
                  sx={{
                    color: "black",
                    fontSize: "0.875rem",
                    fontWeight: "500",
                    textDecoration: "underline",
                    textTransform: "capitalize",
                    "&:hover": {
                      color: "red",
                      textDecoration: "underline",
                    },
                  }}
                  disableRipple
                >
                  {row.subscriptionName}
                </Button>
              </TableCell>

              <TableCell align="center" sx={{ fontSize: "0.875rem" }}>
                {row.countryIds?.length > 1 ? (
                  <Tooltip
                    title={`${row.countryIds?.length} countries are selected`}
                    arrow
                  >
                    <Box>{row.filteredCountryNames?.join(", ")}</Box>
                  </Tooltip>
                ) : (
                  <Tooltip title={"1 country is selected"} arrow>
                    <Box>{row.filteredCountryNames?.join(", ")}</Box>
                  </Tooltip>
                )}
              </TableCell>
              <TableCell align="center" sx={{ fontSize: "0.875rem" }}>
                {row.urgencyArray.join(", ")}
              </TableCell>
              <TableCell align="center" sx={{ fontSize: "0.875rem" }}>
                {row.severityArray.join(", ")}
              </TableCell>
              <TableCell align="center" sx={{ fontSize: "0.875rem" }}>
                {row.certaintyArray.join(", ")}
              </TableCell>
              <TableCell align="center" sx={{ fontSize: "0.875rem" }}>
                {row.subscribeBy.join(", ")}
              </TableCell>
              <TableCell align="center" sx={{ minWidth: "130px" }}>
                {/* <Button
                  variant="text"
                  size="small"
                  color="error"
                  onClick={() => handleOpen("Edit", row.id)}
                  sx={{ minWidth: 0, marginRight: "5px" }}
                >
                  View
                </Button> */}
                <Button
                  variant="text"
                  size="small"
                  onClick={() => handleOpen("Edit", row.id)}
                  sx={{
                    color: "red",
                    minWidth: 0,
                    padding: "5px",
                    mr: "5px",
                    fontSize: "0.875rem",
                    "&:hover": {
                      opacity: "0.7",
                    },
                  }}
                >
                  Edit
                </Button>
                <Button
                  variant="text"
                  size="small"
                  onClick={() => setConfirmDelete(true)}
                  sx={{
                    color: "red",
                    minWidth: 0,
                    padding: "5px",
                    fontSize: "0.875rem",
                    "&:hover": {
                      opacity: "0.7",
                    },
                  }}
                >
                  Delete
                </Button>
                <Dialog
                  open={confirmDelete}
                  onClose={() => {
                    setConfirmDelete(false);
                  }}
                  aria-labelledby="alert-dialog-title"
                  aria-describedby="alert-dialog-description"
                >
                  <DialogTitle id="alert-dialog-title">
                    {"Delete the subscription"}
                  </DialogTitle>
                  <DialogContent>
                    <DialogContentText id="alert-dialog-description">
                      Are you sure to delete this subscription?
                    </DialogContentText>
                  </DialogContent>
                  <DialogActions>
                    <Button
                      onClick={() => {
                        setConfirmDelete(false);
                      }}
                      variant="outlined"
                      color="error"
                      size="small"
                    >
                      Cancel
                    </Button>
                    <Button
                      onClick={() => {
                        handleDelete(row.id);
                        setConfirmDelete(false);
                      }}
                      autoFocus
                      variant="contained"
                      color="error"
                      size="small"
                    >
                      Confirm
                    </Button>
                  </DialogActions>
                </Dialog>
              </TableCell>
            </TableRow>
          ))}
          {emptyRows > 0 && (
            <TableRow
              style={{
                height: 62 * emptyRows,
              }}
            >
              <TableCell colSpan={7} />
            </TableRow>
          )}
        </TableBody>
      </Table>
      {visibleRows?.length === 0 && (
        <Typography
          variant="h6"
          textAlign={"center"}
          padding={"20px"}
          color={"gray"}
        >
          {" "}
          No subscription groups added.
        </Typography>
      )}
      <TablePagination
        rowsPerPageOptions={[5, 10]}
        component="div"
        count={updatedRows ? updatedRows.length : 0}
        rowsPerPage={rowsPerPage}
        page={page}
        onPageChange={handleChangePage}
        onRowsPerPageChange={handleChangeRowsPerPage}
      />
      <ModalForm
        selectedRow={selectedRow}
        setSelectedRow={setSelectedRow}
        formType={formType}
        tableData={tableData}
        setTableData={setTableData}
        countryList={countryList}
        regionList={regionList}
        open={open}
        handleClose={handleClose}
      />
    </TableContainer>
  );
};

export default SubscriptionTable;
