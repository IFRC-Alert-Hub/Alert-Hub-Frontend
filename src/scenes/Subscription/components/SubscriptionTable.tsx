import { useMemo, useState } from "react";
import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Paper,
  TablePagination,
  Button,
  Typography,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogContentText,
  DialogActions,
} from "@mui/material";

import { useMutation } from "@apollo/client";
import { DELETE_SUBSCRIPTION } from "../../../GraphQL API/mutations/subscriptionMutations";
import { subscription_module } from "../../../GraphQL API/API_Links";
import { GET_SUBSCRIPTIONS } from "../../../GraphQL API/ALL_QUERIES";
import {
  SubscriptionForm,
  UpdatedSubscriptionItem,
} from "../../../GraphQL API/TYPES";
import CollapsibleRow from "./CollapsibleRow";

type PropsType = {
  updatedSubscriptions: UpdatedSubscriptionItem[];
  modalOpen: boolean;
  handleModalOpen: () => void;
  setFormType: React.Dispatch<React.SetStateAction<string>>;
  setSelectedRow: React.Dispatch<React.SetStateAction<SubscriptionForm>>;
};

const SubscriptionTable = ({
  updatedSubscriptions,
  modalOpen,
  handleModalOpen,
  setFormType,
  setSelectedRow,
}: PropsType) => {
  const [deleteSubscription] = useMutation(DELETE_SUBSCRIPTION, {
    refetchQueries: [GET_SUBSCRIPTIONS, "FetchSubscriptions"],
    client: subscription_module,
  });
  const [page, setPage] = useState(0);
  const [rowsPerPage, setRowsPerPage] = useState(5);
  const [deleteId, setDeleteId] = useState("");
  const [confirmDelete, setConfirmDelete] = useState(false);

  // Avoid a layout jump when reaching the last page with empty rows.
  const emptyRows = useMemo(
    () =>
      page > 0
        ? Math.max(0, (1 + page) * rowsPerPage - updatedSubscriptions?.length)
        : 0,
    [page, rowsPerPage, updatedSubscriptions]
  );
  const visibleRows = useMemo(
    () =>
      updatedSubscriptions?.slice(
        page * rowsPerPage,
        page * rowsPerPage + rowsPerPage
      ),
    [page, rowsPerPage, updatedSubscriptions]
  );

  const handleChangePage = (
    event: React.MouseEvent<HTMLButtonElement> | null,
    newPage: number
  ) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setRowsPerPage(parseInt(event.target.value, 10));
    setPage(0);
  };

  const handleDelete = (id: string) => {
    deleteSubscription({ variables: { subscriptionId: parseInt(id) } });
  };

  return (
    <TableContainer component={Paper}>
      <Table sx={{ minWidth: 650 }} aria-label="subscription table">
        <TableHead>
          <TableRow>
            <TableCell />
            <TableCell
              align="left"
              sx={{
                minWidth: "180px",
                fontSize: "0.875rem",
                fontWeight: "600",
              }}
            >
              Subscription Name
            </TableCell>
            <TableCell
              align="left"
              sx={{
                fontSize: "0.875rem",
                fontWeight: "600",
              }}
            >
              Country
            </TableCell>
            <TableCell
              align="left"
              sx={{
                fontSize: "0.875rem",
                fontWeight: "600",
              }}
            >
              Admin1s
            </TableCell>
            <TableCell
              align="left"
              sx={{ fontSize: "0.875rem", fontWeight: "600" }}
            >
              Urgency
            </TableCell>
            <TableCell
              align="left"
              sx={{ fontSize: "0.875rem", fontWeight: "600" }}
            >
              Severity
            </TableCell>
            <TableCell
              align="left"
              sx={{ fontSize: "0.875rem", fontWeight: "600" }}
            >
              Certainty
            </TableCell>
            <TableCell
              align="left"
              sx={{ fontSize: "0.875rem", fontWeight: "600" }}
            >
              Operations
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {visibleRows?.map((row: UpdatedSubscriptionItem) => (
            // <TableRow key={row.id}>

            // </TableRow>
            <CollapsibleRow
              key={row.id}
              row={row}
              handleModalOpen={handleModalOpen}
              setFormType={setFormType}
              setSelectedRow={setSelectedRow}
              setDeleteId={setDeleteId}
              setConfirmDelete={setConfirmDelete}
            />
          ))}
          {emptyRows > 0 && (
            <TableRow
              style={{
                height: 62 * emptyRows,
              }}
            >
              <TableCell colSpan={8} />
            </TableRow>
          )}
        </TableBody>
      </Table>
      {visibleRows?.length === 0 && (
        <Typography
          variant="h6"
          textAlign={"center"}
          padding={"20px"}
          color={"gray"}
        >
          {" "}
          No subscription groups added.
        </Typography>
      )}
      <TablePagination
        rowsPerPageOptions={[5, 10]}
        component="div"
        count={updatedSubscriptions ? updatedSubscriptions.length : 0}
        rowsPerPage={rowsPerPage}
        page={page}
        onPageChange={handleChangePage}
        onRowsPerPageChange={handleChangeRowsPerPage}
      />
      <Dialog
        open={confirmDelete}
        onClose={() => {
          setConfirmDelete(false);
        }}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
        sx={{
          ".MuiDialog-paper": {
            color: "transparent",
          },
        }}
      >
        <DialogTitle id="alert-dialog-title">
          {"Delete the subscription"}
        </DialogTitle>
        <DialogContent>
          <DialogContentText id="alert-dialog-description">
            Are you sure to delete this subscription?
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button
            onClick={() => {
              setConfirmDelete(false);
            }}
            variant="outlined"
            color="error"
            size="small"
          >
            Cancel
          </Button>
          <Button
            onClick={() => {
              handleDelete(deleteId);
              setConfirmDelete(false);
            }}
            autoFocus
            variant="contained"
            color="error"
            size="small"
          >
            Confirm
          </Button>
        </DialogActions>
      </Dialog>
    </TableContainer>
  );
};

export default SubscriptionTable;
