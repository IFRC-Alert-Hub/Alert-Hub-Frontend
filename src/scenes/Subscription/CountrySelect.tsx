import { FormControl, InputAdornment, TextField } from "@mui/material";
import ContinentCollapse from "./ContinentCollapse";
import SearchIcon from "@mui/icons-material/Search";

const CountrySelect = () => {
  return (
    <div>
      <FormControl component="fieldset" sx={{ width: "100%" }}>
        <legend className="subs-form-legend">Countries</legend>
        <TextField
          id="search-bar"
          placeholder="Search countries..."
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                <SearchIcon />
              </InputAdornment>
            ),
          }}
          size="small"
          variant="outlined"
          sx={{ width: "100%", borderRadius: 5, mt: 1, mb: 1 }}
        />
        <ContinentCollapse continent="Africa" />
        <ContinentCollapse continent="America" />
        <ContinentCollapse continent="Asia Pacific" />
        <ContinentCollapse continent="Europe" />
        <ContinentCollapse continent="Middle East & North East" />
      </FormControl>
    </div>
  );
};

export default CountrySelect;
