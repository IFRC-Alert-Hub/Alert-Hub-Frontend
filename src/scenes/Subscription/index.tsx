import { Button, Container, Grid, Typography } from "@mui/material";
import { useEffect, useState } from "react";
import { useQuery } from "@apollo/client";
import axios from "axios";
import { subscription_module } from "../../API/API_Links";
import {
  CountryOptionsType,
  SubscriptionForm,
  SubscriptionItem,
  SubscriptionQueryResult,
} from "../../API/TYPES";
import { GET_SUBSCRIPTIONS } from "../../API/ALL_QUERIES";
import SubscriptionTable from "./components/SubscriptionTable";
import ModalForm from "./components/ModalForm";
import Progress from "../../components/Layout/Progress";
import { RADIO_OPTIONS } from "./components/SentFlagRadio";

const INIT_ROW: SubscriptionForm = {
  subscriptionName: "",
  countryIds: [],
  admin1Ids: [],
  urgencyArray: [],
  severityArray: [],
  certaintyArray: [],
  subscribeBy: [],
  sentFlag: -1,
};

const Subscription = () => {
  const [formType, setFormType] = useState("");
  const [modalOpen, setModalOpen] = useState(false);
  const [selectedRow, setSelectedRow] = useState<SubscriptionForm>(INIT_ROW);
  const [countryData, setCountryData] = useState<CountryOptionsType[] | null>(
    null
  );

  const {
    loading: subscriptionLoading,
    error: subscriptionError,
    data: subscriptionData,
  } = useQuery<SubscriptionQueryResult>(GET_SUBSCRIPTIONS, {
    client: subscription_module,
  });

  useEffect(() => {
    axios
      .get("https://alert-manager.azurewebsites.net/admin1s/")
      .then((res) => {
        setCountryData(res.data.countries);
      })
      .catch((error) => {
        console.log(error);
      });
  }, []);

  const handleModalOpen = () => {
    setModalOpen(true);
  };
  const handleModalClose = () => {
    setFormType("");
    setSelectedRow(INIT_ROW);
    setModalOpen(false);
  };

  let tableContent;
  if (subscriptionLoading) {
    tableContent = <Progress />;
  } else if (subscriptionError) {
    tableContent = (
      <Typography variant="h5" textAlign={"center"} color={"gray"} mt={10}>
        Something error! Please contact the application administrator.
      </Typography>
    );
  } else if (subscriptionData && countryData && countryData.length > 0) {
    const tableDetail = subscriptionData.listAllSubscription.map(
      (item: SubscriptionItem) => {
        const countryNames = item.countryIds.map((id: number) => {
          const foundCountry = countryData.find((country) => country.id === id);
          return foundCountry?.name;
        });
        const admin1Names = item.admin1Ids.map((id: number) => {
          const foundCountry = countryData.find(
            (country) => country.id === item.countryIds[0]
          );
          const admin1sList = foundCountry?.admin1s;
          const foundAdmin1s = admin1sList?.find(
            (admin1: any) => admin1.id === id
          );
          return foundAdmin1s?.name;
        });
        const sentFlagName = RADIO_OPTIONS.find(
          (option) => Number(option.value) === item.sentFlag
        )?.label;
        return { ...item, countryNames, admin1Names, sentFlagName };
      }
    );
    tableContent = (
      <SubscriptionTable
        updatedSubscriptions={tableDetail}
        modalOpen={modalOpen}
        handleModalOpen={handleModalOpen}
        setFormType={setFormType}
        setSelectedRow={setSelectedRow}
      />
    );
  } else {
    tableContent = (
      <Typography variant="h5" textAlign={"center"} color={"gray"} mt={10}>
        Something error! Please contact the application administrator.
      </Typography>
    );
  }

  return (
    <Container maxWidth={"lg"}>
      <Grid container direction="row" margin="3rem 0 1rem 0">
        <Grid item xs={12} sm={8}>
          <Typography
            variant="h3"
            fontWeight={"bold"}
            paddingBottom={"15px"}
            textTransform={"capitalize"}
            letterSpacing={"1.6px"}
          >
            My Subscriptions
          </Typography>
        </Grid>
        <Grid item xs={12} sm={4} className="subscription-add-btn">
          <Button
            variant="contained"
            color="error"
            onClick={() => {
              handleModalOpen();
              setFormType("Add");
            }}
            sx={{
              borderRadius: 10,
              marginRight: "10px",
              textTransform: "capitalize",
            }}
          >
            Add
          </Button>
        </Grid>
      </Grid>
      {tableContent}
      {countryData && countryData.length > 0 && (
        <ModalForm
          modalOpen={modalOpen}
          handleModalClose={handleModalClose}
          formType={formType}
          countryList={countryData}
          selectedRow={selectedRow}
          setSelectedRow={setSelectedRow}
        />
      )}
    </Container>
  );
};

export default Subscription;
