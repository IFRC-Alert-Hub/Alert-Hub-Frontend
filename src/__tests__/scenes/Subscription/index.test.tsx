import { act, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Subscription from "../../../scenes/Subscription";
import SubscriptionTable from "../../../scenes/Subscription/components/SubscriptionTable";

describe("Subscription", () => {
  it("render Subscription component", () => {
    render(<Subscription />);

    expect(screen.getByText("My Subscriptions")).toBeInTheDocument();
    expect(screen.getByText("Add")).toBeInTheDocument();
  });
});
