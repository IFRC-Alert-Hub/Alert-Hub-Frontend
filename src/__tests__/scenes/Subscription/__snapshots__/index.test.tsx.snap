// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscirption Snapshot should match DOM snapshot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
>
  <div
    className="MuiGrid-root MuiGrid-container css-ef4x7w-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-8 css-o4bg9w-MuiGrid-root"
    >
      <h3
        className="MuiTypography-root MuiTypography-h3 css-169jn1r-MuiTypography-root"
      >
        My Subscriptions
      </h3>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 subscription-add-btn css-wg4xta-MuiGrid-root"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium css-5acd4l-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Add
      </button>
    </div>
  </div>
  <div
    className="MuiBox-root css-bf5v5p"
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorError css-1lxih1g-MuiCircularProgress-root"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </span>
  </div>
</div>
`;
